Design/WorkspacePropertyWithIndex.md

```diff
@@ -0,0 +1,130 @@
+ - Consistent and simple way of defining index properties without code duplication.
+ - Consistent user interface.
+ 
+## Implementation
+
+### Basics
+
+- Extend the `WorkspaceProperty` type.
+- Own an `ArrayProperty` which can make use of `ListPropertyWidget` for capturing list data from the GUI.
+- Add a set of flags which control allowed index types (for use in the constructor).
+- Use a `PropertyWithValue<std::string>` to capture the selected index type.
+
+### Details
+
+1. The `WorkspacePropertyWithIndex` type constructor will accept the following flags (or combinations of flags):
+	- `IndexType::SpectrumNumber` for list of global<sup>1</sup> spectrum numbers typically from 1-N.
+	- `IndexType::WorkspaceIndex` for list of global workspace indices typically from 0-N.
+	- `IndexType::DetectorID` for list of detector ids within the instrument.
+2. Users will be restricted to entering only one of the allowed index types at the GUI level. 
+3. Validation will occur in `IndexInfo::makeIndexSet`(which throws for bad indices), not in every algorithm.
+4. The `ArrayProperty` will be exposed e.g `WorkspacePropertyWithIndex::getArrayProperty` in order to make use of `ListPropertyWidget`.
+5. Overload function call operator to return `std::pair`.
+6. `IndexInfo` will be used to translate the indices at the point of access i.e. `WorkspacePropertyWithIndex::operator()()`.
+7. There must be a few modifications to the GUI generation code in order to support this type:
+	-  `PropertyWidgetFactory` may need to handle `WorkspacePropertyWithIndex` specifically.
+	-  May need to introduce a concept of a group of property widgets.
+
+<sup>1</sup><sub>The distinction is made here between local (a subset of all data partitioned on an MPI Rank) and global (entire data set). This will not be visible at the GUI or client API level.</sub>
+
+### Example usage
+
+**Property Declaration**:
+
+Declaring the property in C++:
+```cpp
+//Property name may be fixed?
+declareProperty(make_unique<WorkspacePropertyWithIndex<MatrixWorkspace>>("InputWorkspaceWithIndex", IndexType::SpectrumNumber|IndexType::WorkspaceIndex));
+```  
+
+**Algorithm Dialog Box (GUI)**:
+
+**InputWorkspace** [ _________________ ]<br>
+( **.** ) **SpectrumNumber** (  ) **WorkspaceIndex**<br>
+**Indices** [ _________________ ]
+ 
+**Client Code Access**:
+
+*C++*:
+
+Setting the property manually (tuple is created internally):
+
+```cpp
+setProperty("InputWorkspaceWithIndex", ws, std::vector<SpectrumNumber>{1, 2, 3, 4});
+//or
+setProperty("InputWorkspaceWithIndex", ws, IndexType::SpectrumNumber, "1:33,42");
+```
+
+Accessing the property:
+
+```cpp
+MatrixWorkspace_const_sptr inputWS;
+IndexSet indices;
+std::tie(inputWS, indices) = getProperty("InputWorkspaceWithIndex");//returns tuple
+```
+
+*Python*:
+
+The simplest scenario for a call to an algorithm could be:
+```python
+ChangePulsetime(TimeOffset=10, InputWorkspaceWithIndex=("ws", SpectrumNumber, [1:33, 42]))
```

Comment:
I don't think `[1:33, 42]` is valid syntax without an object to slice on.

I'm not too sure about the general form here - it's a lot for users to remember to type. I don't think there's any autocomplete help we could add here. I think this is the most difficult part though.

What about always generating the same available keyword arguments in Python and then combining them internall into the WorkspacePropertyWithIndex, e.g

```python
ChangePulseTime(TimeOffset=10, InputWorkspace=ws, Spectra=("1:33") )
ChangePulseTime(TimeOffset=10, InputWorkspace=ws, Indices=("1:33") )
ChangePulseTime(TimeOffset=10, InputWorkspace=ws, IDs=("1:33") )
```
In this case the default of leaving out the `Spectra/Indices/IDs` just applies it to the whole workspace as now? 

Is this too ambiguous for people?